armi/reactor/converters/uniformMesh.py

```diff
@@ -269,32 +286,254 @@ def makeAssemWithUniformMesh(sourceAssem, newMesh):
+
+            # Check that the parameters on the destination block have been cleared first before attempting to
+            # map the data. These parameters should be cleared using ``UniformMeshGeometryConverter.clearStateOnAssemblies``.
+            for getter, paramNames in [
+                (BlockParamMapper.scalarParamGetter, blockScalarParamNames),
+                (BlockParamMapper.arrayParamGetter, blockArrayParamNames),
+            ]:
+
+                existingDestBlockParamVals = getter(destBlock, paramNames)
+                clearedValues = [
+                    True if not val else False for val in existingDestBlockParamVals
+                ]
+                if not all(clearedValues):
+                    raise ValueError(
+                        f"The state of {destBlock} on {destinationAssembly} "
+                        f"was not cleared prior to calling ``setAssemblyStateFromOverlaps``. "
+                        f"This is an implementation bug in the mesh converter that should "
+                        f"be reported to the developers. The following parameters should be cleared:\n"
+                        f"Parameters: {paramNames}\n"
+                        f"Values: {existingDestBlockParamVals}"
+                    )
+
+        # The destination assembly is the assembly that the results are being mapped to
+        # whereas the source assembly is the assembly that is from the uniform model. This
+        # loop iterates over each block in the destination assembly and determines the mesh
+        # coordinates that the uniform mesh (source assembly) will be mapped to.
+        for destBlock in destinationAssembly:
+
+            zLower = destBlock.p.zbottom
+            zUpper = destBlock.p.ztop
+            destinationBlockHeight = destBlock.getHeight()
+            # Determine which blocks in the uniform mesh source assembly are
+            # within the lower and upper bounds of the destination block.
+            sourceBlocksInfo = sourceAssembly.getBlocksBetweenElevations(zLower, zUpper)
+
+            if not sourceBlocksInfo:
+                raise ValueError(
+                    f"An error occurred when attempting to map to the "
+                    f"results from {sourceAssembly} to {destinationAssembly}. "
+                    f"No blocks in {sourceAssembly} exist between the axial "
+                    f"elevations of {zLower:<8.3e} cm and {zUpper:<8.3e} cm. "
+                    f"This a major bug in the uniform mesh converter that should "
+                    f"be reported to the developers."
+                )
+
+            # Iterate over each of the blocks that were found in the uniform mesh
+            # source assembly within the lower and upper bounds of the destination
+            # block and perform the parameter mapping.
+            updatedDestVals = collections.defaultdict(float)
+
+            _setNumberDensitiesFromOverlaps(destBlock, sourceBlocksInfo)
```

Comment:
This is new now. Previously this was done outside of this mapping routine, but it's easier if this just does it all IMO.